packages/compiler/core/checker.ts

```diff
@@ -2136,12 +2136,8 @@ export function createChecker(program: Program): Checker {
       type.namespace?.models.set(type.name, type);
     }
 
-    const inheritedPropNames = new Set(
-      Array.from(walkPropertiesInherited(type)).map((v) => v.name)
```

Comment:
(Not related to wanting this particular line of code changed.)

I think walkPropertiesInherited may still need to be changed. I don't expect that callers are prepared to see something and its override(s) returned. I think it should probably not return properties that are overridden, at least not by default. I could see an opt-in arg to include them so that you signal that you understand that.

Here's one example: https://github.com/microsoft/cadl/blob/c5a2a75fed07dd0aa3078b244749331a7b7adccc/packages/rest/src/http/metadata.ts#L156

Looks like if you have, for example if you have a base property marked `@header` and an override that isn't marked as `@header` you will still get the header defined by the base, which seems wrong.


There's also this where I anticipated an issue in getEffectiveType once overrides were possible:

https://github.com/microsoft/cadl/blob/c5a2a75fed07dd0aa3078b244749331a7b7adccc/packages/compiler/core/checker.ts#L4577-L4580

It also looks like member references might do the wrong thing in lateBindMembers by binding inherited over top of overridden. However, I couldn't repro this in the playground, so there's probably something wrong with my reasoning and I'm investigating that.

Reply:
lateBindMembers is working nearly by accident. It goes to replace the override with the base property but that just adds to a duplicate list that is never reported as a diagnostic since it happens in late binding. This adds to my intuition that walkPropertiesInherited should be changed to exlude base properties that are overridden. And it would be great to try to add test cases that have overrides that cover paths through walkPropertiesInherited.

Reply:
Yes, `walkInheritedProperties` didn't work at all like I expected it to, which is why I stopped using it. In the case of A extends B extends C, it returns A.prop when called on C even if B also overrides prop. The current logic I implemented simply traverses the inheritance chain and finds the first override for a given property, returning that. 

I don't understand any of the late binding logic, now what the purpose of walkInhertiedProperties is supposed to be, so I left all of that alone.

Reply:
walkInheritedProperties simply enumerates all of the properties in the inheritance chain (with most derived enumerated before least derived). It is used whenever something wants to loop over all properties of a model including inherited properties. Until this change, overrides were impossible so the implementation was trivial. Now we have to decide what it should do when there are overrides. It's not enough that you could implement this change withou walkInheritedProperties, we have to consider all the other existing callers to walkInheritedProperties. And then we need to either change walkInheritedProperties or change the callers if overrides cause them issues.

Looking at just a few calls to walkInheritedProperties, I can find bugs:

1. Override a `@header` with a non-header and the base header remains (could also be `@path`, `@query`, etc:

https://cadlplayground.z22.web.core.windows.net/prs/1248/?c=aW1wb3J0ICJAY2FkbC1sYW5nL3Jlc3QiOwoKQHNlcnZpY2UoewogIHRpdGxlOiAiV2lkZ2V0IFPGGyIsCn0pCm5hbWVzcGFjZSBEZW1vxxs7Cgp1c2luZyBDYWRsLkh0dHA7Cgptb2RlbCBCYXNlIMRVcHJvcDogc3RyaW5nOwp9yCDHZGV4dGVuZHPQLyJ05gCgfQoKb3AgxA0oKTrINCYge307

2. Cannot intersect or a model with a property override with anything

https://cadlplayground.z22.web.core.windows.net/prs/1248/?c=aW1wb3J0ICJAY2FkbC1sYW5nL3Jlc3QiOwoKQHNlcnZpY2UoewogIHRpdGxlOiAiV2lkZ2V0IFPGGyIsCn0pCm5hbWVzcGFjZSBEZW1vxxs7Cgp1c2luZyBDYWRsLkh0dHA7Cgptb2RlbCBCYXNlIMRVQGhlYWRlciBoMTogc3RyaW5nOwp9yCbHamV4dGVuZHPKNcQtInTmAKR9CgpvcCDEDSgpOscyOw%3D%3D

3. Cannot spread a model a model with a property override

https://cadlplayground.z22.web.core.windows.net/prs/1248/?c=aW1wb3J0ICJAY2FkbC1sYW5nL3Jlc3QiOwoKQHNlcnZpY2UoewogIHRpdGxlOiAiV2lkZ2V0IFPGGyIsCn0pCm5hbWVzcGFjZSBEZW1vxxs7Cgp1c2luZyBDYWRsLkh0dHA7Cgptb2RlbCBCYXNlIMRVQGhlYWRlciBoMTogc3RyaW5nOwp9yCbHamV4dGVuZHPKNcQtInTmAKR9CgpvcCDEDSgpOiB7IC4uLsc3fTs%3D

I suspect there is also a bug in getEffectiveType in the location I mentioned where I say this will need to change once overrides arrive, but (2) and (3) got in the way of reproing it.

lateBindMembers is what makes it possible to reference a member, for example you can reference property P of model M using `P.M`. There isn't an observable bug that I can find there, but if a property is overridden, it is getting lucky that the symbol table is recording a duplicate instead of overwriting.

So far in all these cases, I believe what the caller wants is for the overridden members to be excluded from being enumerated.

Reply:
I'll convert these to test cases.

Reply:
So, regarding item 1, is there a general rule? I don't think we are saying that an overridden property needs to have the same decorators as the base property, or are we? Should this just be a separate linter rule in the Http library?

The other two seem straightforward to me. They should allow the intersection or spreading with the derived property, not the base.

Reply:
I think it would be fine if the header disappeared in (1) which is what would happen if walkInheritedProperties didn't return overridden properties. I believe this same change would fix all 3.

Reply:
I think the general rule is an override completely replaces the base property. If it drops a decorator, then you get the consequences. But now what's happening is both the base and the override are contributing to the swagger and not in any coherent way.

Reply:
Thanks @nguerrera. I added those test cases and updated the logic of walkPropertiesInherited to not return duplicates (it will return the closest relative). Please try out those other scenarios you mentioned and let me know if anything else needs changing.

Reply:
@nguerrera please take a second look.

Reply:
I will look at this first thing in the morning. Sorry that I forgot.

Reply:
I have been looking at this all morning. The effective type issue is not easy to describe or fix. I will share a commit with test cases and the fix when I have it.

Reply:
I sent https://github.com/tjprescott/cadl/pull/1 with the changes I suggest.